core/chain/src/lib.rs

```diff
@@ -58,27 +65,47 @@ where
         *guard.blockchain_storage_mut().best_block_hash().unwrap().unwrap()
     }
 
-    /// Check if block already is known.
-    pub fn is_known(&self, hash: &CryptoHash) -> bool {
+    /// Check if block header already is known.
+    pub fn is_known_header(&self, hash: &CryptoHash) -> bool {
         let mut guard = self.storage.write().expect(POISONED_LOCK_ERR);
         match guard.blockchain_storage_mut().header(hash) {
             Ok(Some(_)) => true,
             _ => false,
         }
     }
 
+    /// Check if block already is known.
+    pub fn is_known_block(&self, hash: &CryptoHash) -> bool {
+        let mut guard = self.storage.write().expect(POISONED_LOCK_ERR);
+        match guard.blockchain_storage_mut().block(hash) {
+            Ok(Some(_)) => true,
+            _ => false,
+        }
+    }
+
+    fn update_best_block_header(&self, header: B::SignedHeader) {
+        let mut guard = self.storage.write().expect(POISONED_LOCK_ERR);
+        guard.blockchain_storage_mut().add_header(header).unwrap();
+    }
+
     fn update_best_block(&self, block: B) {
-        let hash = block.block_hash();
         let mut guard = self.storage.write().expect(POISONED_LOCK_ERR);
-        guard.blockchain_storage_mut().set_best_block_hash(hash).unwrap();
         guard.blockchain_storage_mut().add_block(block).unwrap();
     }
 
+    fn update_existing_header(&self, block: B) {
+        let hash = block.block_hash();
+        let mut guard = self.storage.write().expect(POISONED_LOCK_ERR);
+        guard.blockchain_storage_mut().set_block(&hash, block).unwrap();
+    }
+
     /// Inserts a verified block.
     pub fn insert_block(&self, block: B) {
         let block_hash = block.block_hash();
-        if self.is_known(&block_hash) {
-            // TODO: known header but not known block.
+        if self.is_known_header(&block_hash) {
```

Comment:
`if self.is_known_header(&block_hash) && !self.is_known_block(&block_hash)`. Also why must header be known at this point?

Reply:
it's if is_known and !is_known -> we `update_existing_header`